File: Solutions/122. Best Time to Buy and Sell Stock II/CPP/Best_Time_to_Buy_and_Sell_Stock_II.cpp
#include <algorithm>
#include <iostream>
#include <limits>
#include <vector>

using namespace std;

class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int profit = 0;
        int current_buy = numeric_limits<int>::max();
        for (const auto& p : prices) {
            profit += max(0, p - current_buy);
            current_buy = p;
        }

        return profit;
    }
};

int main() {
    Solution* S = new Solution();

    // input
    vector<int> prices{7, 1, 5, 3, 6, 4};

    cout << S->maxProfit(prices) << endl;

    return 0;
}

File: Solutions/122. Best Time to Buy and Sell Stock II/Python/Best_Time_to_Buy_and_Sell_Stock_II.py
class Solution:
    def maxProfit(self, prices: list[int]) -> int:
        profitAfter_b = float('-inf')
        profitAfter_s = 0

        for p in prices:
            profitAfter_b = max(profitAfter_b, profitAfter_s - p)
            profitAfter_s = max(profitAfter_s, profitAfter_b + p)

        return profitAfter_s

S = Solution()

#input
prices = [7, 1, 5, 3, 6, 4]

print(S.maxProfit(prices))